### PR TITLE
stats: jni layer change

### PR DESCRIPTION
**Pull request by jingwei99**
Description: jni layer change for cardinality support. this pr does not change ios behavior for pulse. 
Risk Level: medium
Testing: ci and manually deployed to lyft ios and android app in local dev environment

### CODE REVIEW

File: library/common/jni/jni_interface.cc [hunk 1]
```diff
@@ -92,6 +92,12 @@ extern "C" JNIEXPORT jint JNICALL Java_io_envoyproxy_envoymobile_engine_JniLibra
     JNIEnv* env,
     jclass, // class
     jlong engine, jstring elements, jint count) {
+  const char** tag_arr[2];
```
Comment (jingwei99) [inline library/common/jni/jni_interface.cc:95]: `tag_arr` and `len` are to be passed to `record_counter_inc` as parameters in [upstream change](https://github.com/envoyproxy/envoy-mobile/blob/dc32ac0aeabd42c2f02d0976ee949a03195f7707/library/common/engine.h#L45)

Comment (snowp) [reply library/common/jni/jni_interface.cc:95]: Just so I understand: how does a local variable declaration of `tags` result in it being populated by a value from Java? Or is this code incomplete?

Comment (jingwei99) [reply library/common/jni/jni_interface.cc:95]: this pr will have to be integrated with a downstream change to have the tags value populated from java

`jobject tags` is to be passed in from downstream java layer as a `List of Tag(String key, String value)`

File: library/common/jni/jni_utility.cc [hunk 2]
```diff
@@ -114,45 +114,51 @@ envoy_data* buffer_to_native_data_ptr(JNIEnv* env, jobject j_data) {
...
+    return nullptr;
+  }
+
+  envoy_headers* native_headers = static_cast<envoy_headers*>(safe_malloc(sizeof(envoy_map_entry)));
+  *native_headers = to_native_headers(env, headers);
+  return native_headers;
+}
+
+envoy_stats_tags to_native_tags(JNIEnv* env, jobjectArray tags) { return to_envoy_data(env, tags); }
+
```
Comment (jingwei99) [inline library/common/jni/jni_utility.cc:136]: this method is not new, it got marked as new bc i made changes to `to_native_headers`

File: library/common/jni/jni_utility.cc [hunk 3]
```diff
@@ -114,45 +114,51 @@ envoy_data* buffer_to_native_data_ptr(JNIEnv* env, jobject j_data) {
...
+envoy_map to_envoy_data(JNIEnv* env, jobjectArray jobjArray) {
```
Comment (goaway) [inline library/common/jni/jni_utility.cc:137]: Let's call this `to_native_map`.

File: library/common/jni/jni_utility.cc [hunk 4]
```diff
@@ -114,45 +114,51 @@ envoy_data* buffer_to_native_data_ptr(JNIEnv* env, jobject j_data) {
...
   // Therefore, the length of the native header array is n envoy_data or n/2 envoy_map_entry.
-  envoy_map_size_t length = env->GetArrayLength(headers);
+  envoy_map_size_t length = env->GetArrayLength(jobjArray);
   if (length == 0) {
-    return envoy_noheaders;
+    return {0, NULL};
   }
 
-  envoy_map_entry* header_array =
+  envoy_map_entry* data_array =
```
Comment (goaway) [inline library/common/jni/jni_utility.cc:145]: How about calling this `entry_array` (and using `entry_key`, `entry_value`)?

File: library/common/jni/jni_utility.cc (hunk 3 above)
Comment (goaway) [reply library/common/jni/jni_utility.cc:137]: And perhaps `s/jobjArray/entries/g`

File: library/common/jni/jni_utility.h [hunk 5]
```diff
@@ -39,3 +39,7 @@ envoy_data* buffer_to_native_data_ptr(JNIEnv* env, jobject j_data);
 envoy_headers to_native_headers(JNIEnv* env, jobjectArray headers);
 
 envoy_headers* to_native_headers_ptr(JNIEnv* env, jobjectArray headers);
+
+envoy_stats_tags to_native_tags(JNIEnv* env, jobjectArray headers);
```
Comment (goaway) [inline library/common/jni/jni_utility.h:43]:
```suggestion
envoy_stats_tags to_native_tags(JNIEnv* env, jobjectArray tags);
```

File: library/common/jni/jni_utility.h [hunk 6]
```diff
@@ -39,3 +39,7 @@ envoy_data* buffer_to_native_data_ptr(JNIEnv* env, jobject j_data);
...
+
+envoy_map to_envoy_data(JNIEnv* env, jobjectArray jobjArray);
```
Comment (goaway) [inline library/common/jni/jni_utility.h:45]:
```suggestion
envoy_map to_native_map(JNIEnv* env, jobjectArray entries);
```

**Comment by goaway** [comment]
Looks good so far, just some comments on naming.

**Pull request merged**